crates/ruma-client-api/src/peeking/listen_to_new_events.rs

```diff
@@ -0,0 +1,90 @@
+//! `GET /_matrix/client/*/events`
+//!
+//! Listen for new events related to a particular room.
+
+pub mod v3 {
+    //! `/v3/` ([spec])
+    //!
+    //! [spec]: https://spec.matrix.org/latest/client-server-api/#peeking_get_matrixclientv3events
+
+    use std::time::Duration;
+
+    use ruma_common::{
+        api::{request, response, Metadata},
+        metadata,
+        serde::Raw,
+        OwnedRoomId,
+    };
+    use ruma_events::AnyTimelineEvent;
+
+    const METADATA: Metadata = metadata! {
+        method: GET,
+        rate_limited: false,
+        authentication: AccessToken,
+        history: {
+            1.0 => "/_matrix/client/r0/events",
+            1.1 => "/_matrix/client/v3/events",
+        }
+    };
+
+    /// Request type for the `listen_to_new_events` endpoint.
+    #[request(error = crate::Error)]
+    #[derive(Default)]
+    pub struct Request {
+        /// The token to stream from.
+        ///
+        /// This token is either from a previous request to this API or from the initial sync API.
+        #[ruma_api(query)]
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub from: Option<String>,
+
+        /// The room ID for which events should be returned.
+        #[ruma_api(query)]
+        #[serde(skip_serializing_if = "Option::is_none")]
+        pub room_id: Option<OwnedRoomId>,
+
+        /// The maximum time to wait for an event.
+        #[ruma_api(query)]
+        #[serde(
+            with = "ruma_common::serde::duration::opt_ms",
+            default,
+            skip_serializing_if = "Option::is_none"
+        )]
+        pub timeout: Option<Duration>,
+    }
+
+    impl Request {
+        /// Creates a `Request` for the given room.
+        pub fn new(room_id: OwnedRoomId) -> Self {
```

Comment:
I find it weird that `room_id` and `from` are optional, because the spec says to call `/initialSync` first, so I decided to require at least `room_id` in the constructor.

The reason is probably because the same endpoint was used instead of `/sync`, where it doesn't require a `room_id` to listen to all rooms.

Synapse requires the `room_id` for guests on this endpoint, which should match the definition of "peeking".

An empty request can always be built with `Request::default()`.

Reply:
I opened a spec issue and added a link to it.